File: Backend/src/Cike.Workflow.Core/Activities/Abstracts/ContainerActivity.cs
using Cike.Workflow.Core.Variables;

namespace Cike.Workflow.Core.Activities.Abstracts;

public abstract class ContainerActivity : Activity, IVariableContainer
{
    public ICollection<IActivity> Activities { get; set; } = new List<IActivity>();

    public ICollection<Variable> Variables { get; set; } = new Collection<Variable>();

    /// <inheritdoc />
    protected override async ValueTask ExecuteAsync(ActivityExecutionContext context)
    {
        // Ensure variables have names.
        EnsureNames(Variables);

        // Register variables.
        context.ExpressionExecutionContext.Memory.Declare(Variables);

        // Schedule children.
        await ScheduleChildrenAsync(context);
    }

    private void EnsureNames(IEnumerable<Variable> variables)
    {
        var count = 0;

        foreach (var variable in variables)
            if (string.IsNullOrWhiteSpace(variable.Name))
                variable.Name = $"Variable{++count}";
    }

    /// <summary>
    /// Schedule the <see cref="Activities"/> for execution.
    /// </summary>
    protected virtual ValueTask ScheduleChildrenAsync(ActivityExecutionContext context)
    {
        return ValueTask.CompletedTask;
    }
}

File: Frontend/cike-workflow/src/core/activities/container-activity.ts
/**
 * 容器活动基类：持有子活动集合与变量集合。
 * 对应后端 Cike.Workflow.Core.Activities.Abstracts.ContainerActivity
 */

import { Activity, type ActivityJson } from './activity'
import { parseActivityList } from './registry'
import { Variable } from '../memory/variable'

export abstract class ContainerActivity extends Activity {
    activities: Activity[] = []
    variables: Variable[] = []

    protected constructor(type: string) {
        super(type)
    }

    protected applyContainerJson(json: ActivityJson): void {
        this.activities = parseActivityList(json.activities)
        this.variables = Variable.parseList(json.variables)
    }

    protected containerJson(): ActivityJson {
        return {
            activities: this.activities.map(a => a.toJson()),
            variables: this.variables.map(v => v.toJson()),
        }
    }

    /** 容器默认把全部子活动纳入校验递归；Flowchart 等可覆写 childActivities */
    protected override childActivities(): Activity[] {
        return this.activities
    }

    /** 按 id 查找直接子活动 */
    findChildById(id: string): Activity | undefined {
        return this.activities.find(a => a.id === id)
    }
}
